aea/helpers/dialogue/base.py

```diff
@@ -412,90 +395,145 @@ def last_message(self) -> Optional[Message]:
+        """
+        return message.sender == self.agent_address
+
+    def is_message_by_other(self, message: Message) -> bool:
+        """
+        Check whether the message is by the counterparty agent in this dialogue or not.
+
+        :param message: the message
+        :return: True if message is by the counterparty agent in this dialogue, False otherwise
+        """
+        return not self.is_message_by_self(message)
+
+    def get_message(self, message_id: int) -> Message:
         """
         Get the message whose id is 'message_id'.
 
-        :param message_id_to_find: the id of the message
+        :param message_id: the id of the message
         :return: the message if it exists, None otherwise
         """
-        result = None  # type: Optional[Message]
+        assert not self.is_empty, "This dialogue is empty"
+        assert self.has_message_id(
+            message_id
+        ), "A message with this id does not exists. Expected a message id between 1 and {}. Found {}.".format(
+            self.last_message.message_id, message_id
+        )
+
         list_of_all_messages = self._outgoing_messages + self._incoming_messages
         for message in list_of_all_messages:
-            if message.message_id == message_id_to_find:
-                result = message
-                break
+            if message.message_id == message_id:
+                return message
 
-        return result
+    def has_message(self, message: Message) -> bool:
+        """
+        Check whether a message exists in this dialogue.
 
-    @property
-    def is_empty(self) -> bool:
+        :param message: the message
+        :return: True if message exists in this dialogue, False otherwise
         """
-        Check whether the dialogue is empty.
+        if self.is_empty:
+            return False
 
-        :return: True if empty, False otherwise
+        if not self.has_message_id(message.message_id):
+            return False
+
+        retrieved_message = self.get_message(message.message_id)
+        return message == retrieved_message
+
+    def has_message_id(self, message_id: int) -> bool:
         """
-        return len(self._outgoing_messages) == 0 and len(self._incoming_messages) == 0
+        Check whether a message with the supplied message id exists in this dialogue.
 
-    def update(self, message: Message) -> bool:
+        :param message_id: the message id
+        :return: True if message with that id exists in this dialogue, False otherwise
+        """
+        if self.is_empty:
+            return False
+
+        return 1 <= message_id <= self.last_message.message_id
+
+    def update(self, message: Message) -> None:
         """
         Extend the list of incoming/outgoing messages with 'message', if 'message' belongs to dialogue and is valid.
 
         :param message: a message to be added
         :return: True if message successfully added, false otherwise
```

Comment:
Update docstring return type and mention it can raise `:raises: ...`

Reply:
Done